Update linter versions for `pylint`, `black`, `bandit`

martimlobao opened the pull request:
- update `pylint` to `>=2.8.2,<2.9` from `>=2.4.4,<2.5` (2.4.4 is 19 months behind 2.8.2)
- update `black` to `==21.5b1` from `==20.8b1` (20.8b1 is 9 months behind 21.5b1)
- update `bandit` to `>=1.7.0,<1.8` from `>=1.6.2,<1.7` (1.6.3 was yanked, 1.6.2 is 5 months behind 1.7.0)

Closes #12133

Eric-Arellano reviewed (approved):
Awesome! Thanks :) 

It's plausible the Black upgrade will cause CI to fail. I'll send over a diff if that happens.

martimlobao commented:
@Eric-Arellano Also took the opportunity to update any other linter that happened to be out of date. LMK if you don't want to update those for whatever reason and I'll remove them.

Eric-Arellano commented:
It looks like you hit a Rust test failure that I haven't seen before and likely is a flake 👀

To workaround, you can add a new empty commit with `git commit --allow-empty` and include in the commit message the special strings `[ci skip-rust]` and `[ci skip-build-wheels]`, which will skip some irrelevant CI like Rust tests. Your PR doesn't impact Rust at all, so no need to run those tests.

martimlobao commented:
Sorry, these past couple days have been extremely busy at work, but I'm hoping to come back to this later this week/weekend.

Eric-Arellano commented:
No worries! It was a small change to the commit message to skip a flaky test, so I went ahead and pushed directly. Should go green on this next run and be merged :)

martimlobao commented:
@Eric-Arellano Ok, so I looked into this a little more and I'm not exactly sure what the problem is. My best guess is that something is this is an [issue](https://github.com/pantsbuild/pants/pull/12135/checks?check_run_id=2688573875#step:11:280) with `pkg_resources`, but I'm basically just repeating what the error is saying. I tried running the tests locally but I get even more errors, including `ModuleNotFoundError: No module named 'pint'`.

FYIY, this appears to be related to some change in `pylint` 2.7 (using `version = "pylint>=2.6.2,<2.7"` seems to work fine).

Any pointers on getting this fixed?

Eric-Arellano commented:
> I tried running the tests locally but I get even more errors, including ModuleNotFoundError: No module named 'pint'.

Sorry about that. That should now be fixed. Pull the latest from `main`.

> FYIY, this appears to be related to some change in pylint 2.7 (using version = "pylint>=2.6.2,<2.7" seems to work fine).

Hm, honestly I recommend changing this PR to only upgrade to that version range. We can upgrade to the latest in a dedicated followup. That way we get the benefit of the rest of this PR.

But, otherwise, the path forward is to read Pylint's changelog and try to understand what has changed. For example, perhaps they added new checks that cause tests that used to succeed to now fail?

martimlobao commented:
@Eric-Arellano alright, reverted pylint back to v2.6.2 for now, I'll add a separate ticket to fix the issues with upgrading it to v2.8.2 and try to look into it. Also updated `src/python/pants/backend/python/lint/pylint/rules_integration_test.py` with the correct version of pylint, which I had missed before. Hopefully the tests will pass now when you approve the workflow :)

stuhood commented:
Hey @martimlobao : thanks for the patch! I've kicked off Actions, and will merge on green.

stuhood commented:
Hey @martimlobao : it looks like there is a legit test failure. Mind checking that out?

stuhood commented:
@martimlobao : Mm, this is slightly more challenging to resolve actually! Lemme see.

@Eric-Arellano : My first instinct is that we effectively need to use:
```
bandit>=1.7.0,<1.8; python_version > '3'
bandit==1.6.2; python_version < '3'
```
...as the `PythonToolRequirementsBase.default_version` value, and to then propagate through edits to allow using multiple requirements lines to callers of that class... does that sound right? If so, I can land that as a separate change first.

Eric-Arellano commented:
> My first instinct is that we effectively need to use:

Simpler is to have the default stop working with Python 2 out of the box. We've already done that several times, including with Pytest using Pytest 5+. We no longer claim Pants works out of the box with Python 2 code, only that it is possible to do.

The test would need to be updated accordingly, similar to this test:

https://github.com/pantsbuild/pants/blob/1ea924e38b9eb04e8384ad8b2bc4f94caf60837b/src/python/pants/backend/python/lint/pylint/rules_integration_test.py#L121-L124

stuhood commented:
Thanks for the contribution @martimlobao : sorry for the huge delay.

martimlobao commented:
Thanks @stuhood, glad I was able to submit my first (very small) contribution! I'll begin taking a look at #12188 when I get a little time. 🙂

Eric-Arellano commented:
Thanks @martimlobao! Made it in time for the first 2.6 release candidate
